checker/src/org/checkerframework/checker/lock/LockVisitor.java

```diff
@@ -353,9 +576,319 @@ protected boolean skipContractCheck(Tree tree, FlowExpressions.Receiver expr, Fl
                 }
             }
         } catch (FlowExpressionParseException e) {
-            // errors are reported at declaration site
+            checker.report(e.getResult(), tree);
         }
 
         return false;
     }
+
+    @Override
+    public boolean isValidUse(AnnotatedDeclaredType declarationType,
+            AnnotatedDeclaredType useType, Tree tree) {
+        declarationType.replaceAnnotation(GUARDEDBY);
+        useType.replaceAnnotation(GUARDEDBY);
+
+        return super.isValidUse(declarationType, useType, tree);
+    }
+
+    // When visiting a method invocation, issue an error if the side effect annotation
+    // on the called method causes the side effect guarantee of the enclosing method
+    // to be violated. For example, a method annotated with @ReleasesNoLocks may not
+    // call a method annotated with @MayReleaseLocks.
+    // Also check that matching @GuardSatisfied(index) on a method's formal receiver/parameters matches
+    // those in corresponding locations on the method call site.
+    @Override
+    public Void visitMethodInvocation(MethodInvocationTree node, Void p) {
+
+        SideEffectAnnotation seaOfInvokedMethod = atypeFactory.methodSideEffectAnnotation(TreeUtils.elementFromUse(node), false);
+
+        MethodTree enclosingMethod = TreeUtils.enclosingMethod(atypeFactory.getPath(node));
+
+        ExecutableElement methodElement = null;
+        if (enclosingMethod != null) {
+            methodElement = TreeUtils.elementFromDeclaration(enclosingMethod);
+        }
+
+        SideEffectAnnotation seaOfContainingMethod = atypeFactory.methodSideEffectAnnotation(methodElement, false);
+
+        if (seaOfInvokedMethod.isWeakerThan(seaOfContainingMethod)) {
+            checker.report(Result.failure(
+                    "method.guarantee.violated",
+                    seaOfContainingMethod.getNameOfSideEffectAnnotation(),
+                    methodElement.toString(),
+                    TreeUtils.elementFromUse(node).toString(),
+                    seaOfInvokedMethod.getNameOfSideEffectAnnotation()), node);
+
+            // TODO: issue method.guarantee.violated if a @LockingFree method is synchronized or
```

Comment:
Wouldn't you do this at the method declaration rather than the method invocation? Could you move this todo to the place where the checking would occur?